microcluster/app.go

```diff
@@ -44,9 +44,9 @@ type Args struct {
 	StateDir    string
 	SocketGroup string
 
-	ListenPort string
-	Client     *client.Client
-	Proxy      func(*http.Request) (*url.URL, error)
+	ListenAddress string
```

Comment:
I'm fine with making this a whole address, but can you please rename it to `PreInitListenAddress` to make it clear what it's for?

Also, this is going to break the daemon's authentication because we detect if we are in a pre-init listening mode by checking if the address is blank: https://github.com/canonical/microcluster/blob/fe0e5ecba28a2898ff81bbe71b60e87300b610da/rest/access/handlers.go#L56-L59

Reply:
It's not sufficient to check for the database status because the core API comes up before the database does.

I suppose you can add a helper to `Endpoints` to fetch the certificate used by the core server and check if it's the `server.crt` here.

Reply:
Good catch, thx. Should we check for `state.Database.Status() == StatusNotReady` instead?

Reply:
Nvm, I saw your other comment too late :)

Reply:
But in between the core API being started and the DB becoming ready, is there any requirement to already authenticate requests?
Just wondering because the check for `StatusNotReady` is "cheap" as it will run every time we authenticate a request.

Reply:
I'm mainly concerned about races and ongoing connections leading to security issues. Think about this:
* You use `PreInitListenAddress` to allow all incoming connections without authentication
* You run `init` to set up your system
* A bad actor queries your cluster on an endpoint with `AllowedBeforeInit=true`, database is offline so it's allowed
* It's an expensive endpoint and takes time to run
* The database comes online
* The request is still ongoing and at this point checks if the database is online, it is so it makes database changes.



Besides, just fetching the cert from `Endpoints` is just as cheap, it's [stored here already](https://github.com/canonical/microcluster/blob/fe0e5ecba28a2898ff81bbe71b60e87300b610da/internal/endpoints/network.go#L20)

Reply:
I see makes sense. I have just pushed the check for the server's certificate. That feels to be the most bullet proof.

I have tested this with the `/cluster/1.0/ready` endpoint as it is allowed before init and doesn't require the DB (which would post the "error not initialized" prior to init).